### PR TITLE
fix: Incomplete json filter paths render errors

### PR DESCRIPTION
**Pull request by robluton**
## What's Changed

- Added `convertToSQLitePath` (`api/src/database/helpers/fn/json/sqlite-json-path.ts`) to build valid SQLite `json_extract` paths from filter/function query paths
- Incomplete paths (e.g. a trailing dot `.author.` left mid-edit in the filter UI) are now normalized to valid paths
- Handles property access, root array indices, mixed nested access, numeric object keys, and bracketed non-numeric segments

## Tested Scenarios

- Manually typed a JSON path character-by-character in the filter UI and confirmed no errors appear

## Checklist

> Leave unchecked where not applicable

- [x] Tests added/updated
- [ ] Documentation PR created in [directus/docs](https://github.com/directus/docs) 
- [ ] OpenAPI updated
  - [ ] Local specs package (`@directus/specs`)
  - [ ] PR created in [directus/openapi](https://github.com/directus/openapi) 
- [ ] SDK (`@directus/sdk`) updated to reflect the changes
- [ ] Types (`@directus/types`) updated to reflect the changes
- [ ] GraphQL schema updated to reflect the changes
- [ ] System data (`@directus/system-data`) updated for changes to system collections/fields/relations
- [ ] Database migration added for schema/system changes
- [ ] Environment variables documented for new/changed config
- [ ] App translations added for new user-facing strings
- [ ] Security implications apply

Fixes [cms-2868](https://linear.app/directus/issue/CMS-2868/incomplete-json-filter-paths-break-the-filter-ui-and-api-requests)

### CODE REVIEW

File: packages/composables/src/use-items.ts [hunk 1]
```diff
@@ -58,6 +58,29 @@ function hasJsonOperator(value: unknown): boolean {
...
+
+function hasIncompleteJsonPathFilter(value: unknown): boolean {
+	if (!value || typeof value !== 'object' || Array.isArray(value)) return false;
+
+	return Object.entries(value).some(([path, child]) => {
+		if (path === '_and' || path === '_or') {
+			return Array.isArray(child) && child.some(hasIncompleteJsonPathFilter);
+		}
+
+		return !/^(?:[\p{L}\p{N}\p{Extended_Pictographic}_]+|\[\d+\])(?:\.[\p{L}\p{N}\p{Extended_Pictographic}_]+|\[\d+\])*$/u.test(
```
Comment (alvarosabu) [inline packages/composables/src/use-items.ts]: Blocking:

I think this regex is stricter than what the API accepts https://github.com/directus/directus/blob/dd03f3df2e38ca7b6ff01cf8a14a910a0ab97efa/api/src/database/helpers/fn/json/parse-function.ts#L37-L48 so we might be hiding some errors

parseJsonPath on the API side doesn't validate path structure. It only bans a few substrings `([], .., *?@$, [-)` and checks characters. The actual parsing is `lodash` toPath, which doesn't care whether brackets contain numbers.

An example `data[abc]` is a valid, working filter (->'data'->'abc' on Postgres). This regex marks it incomplete, so we never send the request. The user gets an empty list with no error and no clue why.

Comment (alvarosabu) [reply packages/composables/src/use-items.ts]: Same regex also swallows paths the backend rejects `(.., @/$/*, [-1])`

File: packages/composables/src/use-items.ts
Comment (alvarosabu) [inline packages/composables/src/use-items.ts:1]: Mmm `hasJsonOperator` / `hasIncompleteJsonPath` / `hasIncompleteJsonPathFilter` are three almost-identical recursive walkers.

I would probably replace both new functions with one woalker that yeilds every path key `_json`

```ts
function* jsonFilterPaths(value: unknown, inJsonFilter = false): Generator<string> {
	if (!value || typeof value !== 'object') return;

	if (Array.isArray(value)) {
		for (const child of value) yield* jsonFilterPaths(child, inJsonFilter);
		return;
	}

	for (const [key, child] of Object.entries(value)) {
		if (inJsonFilter && key !== '_and' && key !== '_or') {
			yield key;
		} else {
			yield* jsonFilterPaths(child, inJsonFilter || key === '_json');
		}
	}
}
```

And then.

```ts
function isCompleteJsonPath(path: string): boolean {
	return /^(?:[\p{L}\p{N}\p{Extended_Pictographic}_]+|\[\d+\])(?:\.[\p{L}\p{N}\p{Extended_Pictographic}_]+|\[\d+\])*$/u.test(path);
}

function hasIncompleteJsonPath(filter: unknown): boolean {
	for (const path of jsonFilterPaths(filter)) {
		if (!isCompleteJsonPath(path)) return true;
	}

	return false;
}
```

I'm pretty sure it would work but worth testing it

**Review by alvarosabu — DISMISSED** [review]
Hey @robluton overall solid fix, just some Dry/correctness points to address, let me know if you have any doubt

File: packages/composables/src/use-items.ts (hunk 1 above)
Comment (br41nslug) [reply packages/composables/src/use-items.ts]: `data[abc]` is not an explicitly/documented supported structure, it just happens to work for postgres and i didnt explicitly guard against it but other databases will fail on such invalid paths. I think its fine for the app to consider `data[abc]` as invalid/incomplete.

The regex should be based on our documented path notation https://directus.com/docs/guides/connect/json/advanced-querying#path-notation not on vendor specific details my sanitization/validation function.

File: packages/composables/src/use-items.ts
Comment (br41nslug) [reply packages/composables/src/use-items.ts:1]: ~~Why would we need a full fletched tree walker for the app side? the interface doesnt allow you to nest any filters inside `_json` like `_and/_or` so it can be safely assumed to always have a single path when created in the app component~~

My bad it's already a tree walker for the entire filter, the function naming made me think it was just for the json filter

File: api/src/database/helpers/fn/json/sqlite-json-path.ts [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+			continue;
+		}
+
+		property += character;
+	}
+
+	appendProperty();
+
+	return result;
+}
```
Comment (br41nslug) [inline api/src/database/helpers/fn/json/sqlite-json-path.ts]:
```suggestion
export function convertToSQLitePath(path: string): string {
	const parts = toPath(path.startsWith('.') ? path.slice(1) : path);

	let result = '$';

	for (const part of parts) {
		if (part === '') continue;

		if (/^\d+$/.test(part)) {
			result += `[${part}]`;
		} else {
			result += `.${part}`;
		}
	}

	return result;
}
```
Same as mysql/postgres we should be using lodash toPath here for consistency.
My suggested version is almost identical to `convertToMySQLPath` so we should probably abstract that to a global `convertJsonPath` function that does this for all vendors that dont have their own converter

**Review by br41nslug — CHANGES_REQUESTED** [review]
I dont think this fully solves the underlying issue, it definitely papers over the issue however the current implementation would still error on paths like `a[.b]` (unlike the `toPath` approach)

I would imagine that MSSQL and Oracle have this same issue considering they're also just doing `const jsonPath = '$' + options.jsonPath;` for their sql queries

File: api/src/database/helpers/fn/json/convert-json-path.ts
Comment (br41nslug) [inline api/src/database/helpers/fn/json/convert-json-path.ts:1]: The function exported here is called `convertToJsonPath` so lets name the file something similar like `convert-json-path.ts` or something.

File: api/src/database/helpers/fn/json/mysql-json-path.test.ts [hunk 3]
```diff
@@ -35,40 +30,22 @@ const TEST_CASES = [
...
-	// Edge cases with underscore and special characters
+	// Edge cases with underscore and numbers in property names
 	{ input: '.user_data', expected: '$.user_data', description: 'property with underscore' },
-	{
-		input: '.first_name.last_name',
-		expected: '$.first_name.last_name',
-		description: 'nested properties with underscores',
-	},
-
-	// Edge cases with numbers in property names
```
Comment (br41nslug) [inline api/src/database/helpers/fn/json/mysql-json-path.test.ts:65]: Why is this test coverage removed?

File: api/src/database/helpers/fn/json/dot-notation-path.ts [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+ * accept it: SQLite (`json_extract`), MySQL/MariaDB (`JSON_EXTRACT`), MSSQL
+ * (`JSON_VALUE`/`JSON_QUERY`), and Oracle (`JSON_VALUE`/`JSON_QUERY`).
+ *
+ * @example ".color" → "$.color"
+ * @example ".items[0].name" → "$.items[0].name"
+ * @example ".author." → "$.author"
+ * @example "a[.b]" → "$.a.b"
+ */
+export function convertToJsonPath(path: string): string {
+	const parts = toPath(path.startsWith('.') ? path.slice(1) : path);
```
Comment (br41nslug) [inline api/src/database/helpers/fn/json/dot-notation-path.ts]:
```suggestion
	const parts = toPath(path);
```
I know i already had this in my old function but looking at how `toPath` works we can probably drop the starting dot check, that would create an empty part in the `toPath` result but we are ignoring empty parts anyway.

**Review by br41nslug — CHANGES_REQUESTED** [review]
Looking good so far. We should still add a blackbox test or 2 for these incomplete paths in `tests/blackbox/tests/db/routes/items/json-filter.test.ts` (it looks like some existing json bb tests are failing for mssql)

Also a behavior change worth noting is that `.123` translates to `$[123]` now instead doing an array index instead of the previous object key index. I think that can be considered an inconsistency in the original implementation that was now made consistent across vendors but still good to be aware of. It could be debatable whether this change makes this a `minor` change instead of a `patch`

File: api/src/database/helpers/fn/json/postgres-json-path.test.ts [hunk 5]
```diff
@@ -89,6 +89,10 @@ const TEST_CASES = [
 		bindings: ['first_name', 'last_name'],
 		description: 'nested properties with underscores',
 	},
+
+	// Incomplete paths — empty segments left by a trailing dot are dropped,
+	// consistent with convertToJsonPath used by the other dialects.
```
Comment (br41nslug) [inline api/src/database/helpers/fn/json/postgres-json-path.test.ts]: I dont think we need a comment here, the test "description" should describe its purpose. altough we could leave the first comment line if intended as grouping of all tests for "incomplete paths"

File: tests/blackbox/tests/db/routes/items/json-filter.test.ts [hunk 6]
```diff
@@ -273,6 +273,15 @@ const getSuccessGroups = (localCollectionSuppliers: string): Record<string, Succ
 			expectedNames: ['Delta'],
 		},
 	],
+	'Incomplete path normalization': [
+		{
+			// normalizes: 'color.' → '$.color', matching Alpha the same as a bare 'color'.
+			description: 'trailing dot normalizes and matches (color. → color)',
```
Comment (br41nslug) [inline tests/blackbox/tests/db/routes/items/json-filter.test.ts]: I dont think we need a comment here, the test "description" should describe its purpose

File: tests/blackbox/tests/db/routes/items/json-filter.test.ts [hunk 7]
```diff
@@ -753,6 +762,11 @@ const ERROR_CASES: ErrorCase[] = [
 		// nested paths must be written flat: 'a.b.c[0]', not { 'a.b': { 'c[0]': { _eq: 1 } } }
 		filter: { metadata: { _json: { 'a.b': { 'c[0]': { _eq: 1 } } } } },
 	},
+	{
+		// A repeated dot forms '..' (recursive descent), which is rejected by parseJsonPath.
+		description: 'returns 400 for a repeated dot (settings..theme)',
```
Comment (br41nslug) [inline tests/blackbox/tests/db/routes/items/json-filter.test.ts]: I dont think we need a comment here, the test "description" should describe its purpose

File: api/src/database/helpers/fn/json/postgres-json-path.ts [hunk 8]
```diff
@@ -17,39 +17,38 @@ export function buildPostgresJsonPath(
...
-	const parts = toPath(path.startsWith('.') ? path.slice(1) : path);
+	const parts = toPath(path);
 
 	let template = '';
 	const bindings: (string | number)[] = [];
 
-	for (let i = 0; i < parts.length - 1; i++) {
+	// Index of the final non-empty segment, which takes the asText operator (->>).
+	let lastIndex = parts.length - 1;
+	while (lastIndex >= 0 && parts[lastIndex] === '') lastIndex--;
```
Comment (br41nslug) [inline api/src/database/helpers/fn/json/postgres-json-path.ts]:
```suggestion
	while (lastIndex >= 0 && parts[lastIndex] === '') {
	  lastIndex--;
	}
```

File: api/src/database/helpers/fn/json/postgres-json-path.ts [hunk 9]
```diff
@@ -17,39 +17,38 @@ export function buildPostgresJsonPath(
...
+
+	for (let i = 0; i <= lastIndex; i++) {
 		const part = parts[i]!;
 
+		// Drop empty segments left behind by incomplete paths (leading/trailing/repeated dots),
+		// consistent with convertToJsonPath used by the other dialects.
```
Comment (br41nslug) [inline api/src/database/helpers/fn/json/postgres-json-path.ts]: I dont think this comment adds much value, `if(part === '')` seems pretty self-explanatory as "skips empty parts"

File: api/src/database/helpers/fn/json/convert-json-path.ts [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+ * @example ".author." → "$.author"
+ * @example "a[.b]" → "$.a.b"
+ */
+export function convertToJsonPath(path: string): string {
+	const parts = toPath(path);
+
+	let result = '$';
+
+	for (const part of parts) {
+		// Drop empty segments left behind by incomplete paths (trailing/repeated dots).
```
Comment (br41nslug) [inline api/src/database/helpers/fn/json/convert-json-path.ts]: I think we can drop this comment too for the same reason as in the postgres function

File: api/src/database/helpers/fn/json/postgres-json-path.ts
Comment (br41nslug) [inline api/src/database/helpers/fn/json/postgres-json-path.ts]:
```suggestion
 * @example ".items[0].name" → { template: "->?->0->?", bindings: ["items", 0, "name"] }
```
Update to the existing example because array indexes are inlined

**Review by br41nslug — COMMENTED** [review]
Looking great! a final round of nitpicks mostly on comments

**Review by br41nslug — APPROVED** [review]
